File: src/java/src/main/java/com/octodemo/vuln/QualitySmells.java
package com.octodemo.vuln;

import java.util.ArrayList; // used below
import java.util.List;      // used below
import java.util.Map;       // java/unused-import - unused import

/**
 * Code-quality smells (NOT security) detected by CodeQL `security-and-quality`.
 * Triggers maintainability alerts: unused imports, unused locals,
 * identical comparisons, container-size comparison, and reference equality on strings.
 */
public class QualitySmells {

    public boolean identicalComparison(int x) {
        // java/comparison-of-identical-expressions: always true
        if (x == x) {
            return true;
        }
        return false;
    }

    public int unusedLocal(List<Integer> values) {
        int unused = 42; // java/unused-local-variable
        int total = 0;
        for (int v : values) {
            total += v;
        }
        return total;
    }

    public boolean stringReferenceEquality(String a) {
        String b = "hello";
        // java/reference-equality-of-boxed-types / string == comparison
        return a == b;
    }

    public boolean sizeCheck(List<String> items) {
        // java/inefficient-empty-string-test style: comparing size() instead of isEmpty()
        if (items.size() == 0) {
            return true;
        }
        return false;
    }

    public List<String> makeList() {
        List<String> list = new ArrayList<>();
        list.add("a");
        return list;
    }

    public boolean usesMap(Map<String, String> ignored) {
        return false;
    }
}

File: src/python/quality_smells.py
"""Code-quality smells (NOT security) detected by CodeQL `security-and-quality`.

These trigger maintainability/reliability alerts such as unused imports,
unused variables, identical comparisons, duplicate dict keys, and unreachable code.
"""

import os
import sys  # noqa: F401 - unused import (py/unused-import)


def comparison_using_is(value):
    # py/comparison-using-is: 'is' compared against a literal
    if value is "active":
        return True
    return False


def identical_comparison(x):
    # py/comparison-of-identical-expressions: always true
    if x == x:
        return 1
    return 0


def duplicate_dict_key():
    # py/duplicate-key-dict-literal
    config = {"timeout": 30, "retries": 3, "timeout": 60}
    return config


def unused_local_and_unreachable(items):
    total = 0
    unused = 42  # py/unused-local-variable
    for item in items:
        total += item
    return total
    print("this never runs")  # py/unreachable-statement


def redundant_comparison(flag):
    # py/redundant-comparison: comparing a boolean to True
    if flag == True:
        return "yes"
    return "no"
